include/nbl/asset/IMeshPacker.h

```diff
@@ -305,29 +258,119 @@ class IMeshPacker : public MeshPackerBase
+                }
+            }
+
+            output[i].triangles = core::vector<Triangle>(m_maxTriangleCountPerMDIData);
+        }
+
+        //struct TriangleMortonCodePair
+        //{
+        //	Triangle triangle;
+        //	//uint64_t mortonCode; TODO after benchmarks
+        //};
+
+        //TODO: triangle reordering
 
-    uint32_t m_vtxSize;
-    uint32_t m_perInstVtxSize;
+        auto* srcIdxBuffer = meshBuffer.getIndexBufferBinding();
+        uint32_t* idxBufferPtr32Bit = static_cast<uint32_t*>(srcIdxBuffer->buffer->getPointer()) + (srcIdxBuffer->offset / sizeof(uint32_t)); //will be changed after benchmarks
+        uint16_t* idxBufferPtr16Bit = static_cast<uint16_t*>(srcIdxBuffer->buffer->getPointer()) + (srcIdxBuffer->offset / sizeof(uint16_t));
+        for (TriangleBatch& batch : output)
+        {
+            for (Triangle& tri : batch.triangles)
+            {
+                if (meshBuffer.getIndexType() == EIT_32BIT)
+                {
+                    tri.oldIndices[0] = *idxBufferPtr32Bit;
+                    tri.oldIndices[1] = *(++idxBufferPtr32Bit);
+                    tri.oldIndices[2] = *(++idxBufferPtr32Bit);
+                    idxBufferPtr32Bit++;
+                }
+                else if (meshBuffer.getIndexType() == EIT_16BIT)
+                {
+
+                    tri.oldIndices[0] = *idxBufferPtr16Bit;
+                    tri.oldIndices[1] = *(++idxBufferPtr16Bit);
+                    tri.oldIndices[2] = *(++idxBufferPtr16Bit);
+                    idxBufferPtr16Bit++;
+                }
+            }
+        }
+
+        return output;
+    }
+
+    core::unordered_map<uint32_t, uint16_t> constructNewIndicesFromTriangleBatch(TriangleBatch& batch, uint16_t*& indexBuffPtr)
+    {
+        core::unordered_map<uint32_t, uint16_t> usedVertices;
+        core::vector<Triangle> newIdxTris = batch.triangles;
+
+        uint32_t newIdx = 0u;
+        for (uint32_t i = 0u; i < batch.triangles.size(); i++)
+        {
+            const Triangle& triangle = batch.triangles[i];
+            for (int32_t j = 0; j < 3; j++)
+            {
+                const uint32_t oldIndex = triangle.oldIndices[j];
+                auto result = usedVertices.insert(std::make_pair(oldIndex, newIdx));
+
+                newIdxTris[i].oldIndices[j] = result.second ? newIdx++ : result.first->second;
+            }
+        }
+
+        //TODO: cache optimization
 
+        //copy indices into unified index buffer
+        for (size_t i = 0; i < batch.triangles.size(); i++)
+        {
+            for (int j = 0; j < 3; j++)
+            {
+                *indexBuffPtr = newIdxTris[i].oldIndices[j];
+                indexBuffPtr++;
+            }
+        }
+
+        return usedVertices;
+    }
+
+    void deinterleaveAndCopyAttribute(MeshBufferType* meshBuffer, uint16_t attrLocation, core::unordered_map<uint32_t, uint16_t>& usedVertices, uint8_t* dstAttrPtr, bool roundUpAttrSizeToPoT /*TODO*/)
```

Comment:
what is roundUpAttrSizeToPoT ?

Reply:
also make the reference to `usedVertices` a constant reference

Reply:
> what is roundUpAttrSizeToPoT ?

irrelevant, I will delete it